compiler/rustc_typeck/src/check/method/mod.rs

```diff
@@ -317,6 +317,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let trait_ref = ty::TraitRef::new(trait_def_id, substs);
 
         // Construct an obligation
+        // here is where we trigger the debug_assertion
+        if trait_ref.has_escaping_bound_vars() {
```

Comment:
this doesn't quite seem like the best solution to me.

We really shouldn't have unbound variables here, so we probably want to move that check further out and if it's intended that either the `self_ty` or the `opt_input_types` can contain bound vars we should use a `Binder<Ty<'tcx>>` instead of a plain ty. cc @jackh726 @nikomatsakis

Reply:
This indeed isn't the right fix. There *can* be late bound vars here. The origin of us "losing" binders is here: https://github.com/rust-lang/rust/blob/1b6b06a03a00a7c9f156bff130b72e90b79e1127/compiler/rustc_typeck/src/check/op.rs#L507
then
https://github.com/rust-lang/rust/blob/1b6b06a03a00a7c9f156bff130b72e90b79e1127/compiler/rustc_typeck/src/check/op.rs#L520

It's a little weird because the code around here is just assuming we can skip binders since these are just suggestions, which isn't the case.

Reply:
Thanks! Now bubbling the check up to `add_type_neq_err_label`.